datavec/datavec-python/src/main/java/org/datavec/python/PythonObject.java

```diff
@@ -77,79 +78,58 @@ public PythonObject(BytePointer bp){
+            case LONG:
+                numpyType = NPY_INT64;
                 break;
             case UINT16:
-                ctype = PyObject_GetAttrString(ctypes, "c_uint16");
+                numpyType = NPY_USHORT;
                 break;
             case UINT32:
-                ctype = PyObject_GetAttrString(ctypes, "c_uint32");
+                numpyType = NPY_UINT;
                 break;
             case UINT64:
-                ctype = PyObject_GetAttrString(ctypes, "c_uint64");
+                numpyType = NPY_UINT64;
                 break;
             case BOOL:
-                ctype = PyObject_GetAttrString(ctypes, "c_bool");
+                numpyType = NPY_BOOL;
                 break;
             case BYTE:
-                ctype = PyObject_GetAttrString(ctypes, "c_byte");
+                numpyType = NPY_BYTE;
                 break;
             case UBYTE:
-                ctype = PyObject_GetAttrString(ctypes, "c_ubyte");
+                numpyType = NPY_UBYTE;
+                break;
+            case HALF:
+                numpyType = NPY_HALF;
                 break;
             default:
                 throw new RuntimeException("Unsupported dtype: " + npArray.getDtype());
         }
 
-        PyObject ctypesPointer = PyObject_GetAttrString(ctypes, "POINTER");
-        PyObject argsTuple = PyTuple_New(1);
-        PyTuple_SetItem(argsTuple, 0, ctype);
-        PyObject ptrType = PyObject_Call(ctypesPointer, argsTuple, null);
-
-        PyObject cast = PyObject_GetAttrString(ctypes, "cast");
-        PyObject address = PyLong_FromLong(npArray.getAddress());
-        PyObject argsTuple2 = PyTuple_New(2);
-        PyTuple_SetItem(argsTuple2, 0, address);
-        PyTuple_SetItem(argsTuple2, 1, ptrType);
-        PyObject ptr = PyObject_Call(cast, argsTuple2, null);
-        PyObject shapeTuple = PyTuple_New(npArray.getShape().length);
-        for (int i = 0; i < npArray.getShape().length; i++) {
-            PyObject dim = PyLong_FromLong(npArray.getShape()[i]);
-            PyTuple_SetItem(shapeTuple, i, dim);
-            Py_DecRef(dim);
-        }
-        PyObject ctypesLib = PyObject_GetAttrString(np, "ctypeslib");
-        PyObject asArray = PyObject_GetAttrString(ctypesLib, "as_array");
-        PyObject argsTuple3 = PyTuple_New(2);
-        PyTuple_SetItem(argsTuple3, 0, ptr);
-        PyTuple_SetItem(argsTuple3, 1, shapeTuple);
-        nativePythonObject = PyObject_Call(asArray, argsTuple3, null);
-
-        Py_DecRef(ctypesPointer);
-        Py_DecRef(ctypesLib);
-        Py_DecRef(argsTuple);
-        Py_DecRef(argsTuple2);
-        Py_DecRef(argsTuple3);
-        Py_DecRef(cast);
-        Py_DecRef(asArray);
-
+        long[] shape = indArray.shape();
+        nativePythonObject = PyArray_New(PyArray_Type(), shape.length, new SizeTPointer(shape),
+                numpyType, null,
+                (dataType == DataType.BFLOAT16 ? indArray.castTo(DataType.FLOAT) : indArray)
```

Comment:
Since the .castTo creates a copy of the array, that should probably raise a warning.

Reply:
Done! I've added a warning when we're casting BFLOAT16 to FLOAT. It looks like this: 
```
21:52:06.037 [main] DEBUG org.datavec.python.PythonGIL - CPython: PyEval_RestoreThread()
21:52:06.037 [main] DEBUG org.datavec.python.PythonGIL - CPython: PyThreadState_Swap()
21:52:06.038 [main] WARN org.datavec.python.PythonObject - 

The given nd4j array 

[[[    1.0000,    1.0000], 
  [    1.0000,    1.0000]], 

 [[    1.0000,    1.0000], 
  [    1.0000,    1.0000]]]

 is of BFLOAT16 datatype. Casting a copy of it to FLOAT and creating the respective numpy array from it.

21:52:06.058 [main] DEBUG org.datavec.python.PythonExecutioner - import sys
import traceback
import json
import inspect
```

Reply:
https://github.com/KonduitAI/deeplearning4j/pull/241/commits/d0c112e25abf18fa36fd872f0c2cdaa3c73a3d3e